Add support for JSON value creation

mransan opened the pull request:
This PR adds the ability to construct JSON value safely and efficiently in OCaml leveraging the JS runtime.

As a side note, these functionality is necessary for the support of JSON serialization in [ocaml-protoc](https://github.com/mransan/ocaml-protoc/)

Signed-Off-By: Maxime Ransan <maxime.ransan@gmail.com>

glennsl commented on jscomp/others/js_dict.ml:
```diff
@@ -27,6 +27,7 @@ type 'a t
 type key = string
 
 external get : 'a t -> key -> 'a option = "" [@@bs.get_index] [@@bs.return {undefined_to_opt}]
+external unsafe_get : 'a t -> key -> 'a = "" [@@bs.get_index] 
```
I'm a bit iffy about this. This isn't really an "unsafe" function, it's just incorrect. While it's certainly not safe, "unsafe" to me means it will throw, not ignore errors and pass on the hot potatoes. Such functions should not exist at all IMO, and I think the correct thing to do here is to define a normal function that uses `get`, but raises an exception on `None`.

mransan replied:
This API is consistent (both name and behavior) with OCaml other `unsafe` function such as `Array.unsafe_get`. Those functions are provided for performance reason when the caller has already checked for the conditions. 

In my use case the caller code, first calls `keys` function and then iterates over those keys. Therefore the check is already done and it is indeed safe to call `unsafe_get`. 

If we want a function which throws exception, then usually the convention in the OCaml community would be to have `exn_get` function. This one will do boundary checks and throws invalid_args if condition is not met.

glennsl replied:
Calling `Array.unsafe_get` in debug mode in OCaml does not cause an out of bounds exception? I find that quite horrifying if true. It's pretty much asking for heartbleed bugs.

glennsl commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +59,29 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+let boolean_array a : t = Obj.magic (Array.map boolean a)
+
+external to_string : 'a -> string = "JSON.stringify" [@@bs.val]
```
Wouldn't it be better to just call this `stringify`?

mransan replied:
Yes I have no problem with this.

mransan commented:
@glennsl I'll make the change to follow JS (camel case) convention and your Doc example is awesome and I will use it.

mransan commented:
@glennsl I could not change the case of `reify_type` since I think @chenglou is using it!

mransan commented:
@glennsl I am also adding an example for parsing and using `reify_type`

glennsl commented:
@mransan You can rename it and just add a deprecated alias, can't you?

```ml
let reifyType ...

let reify_type = reifyType
[@@ocaml.deprecated "Please use `reifyType`"]
```

That way we can at least annoy him a bit :)

mransan commented:
@glennsl @chenglou Yep it's in! `reify_type` is deprecated.

mransan commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +59,29 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+let null : t = Obj.magic Js.null 
+
+let string s : t = Obj.magic s
```
something like this:
```
external string : string -> t = "%identity" 
```

bobzhang replied:
yes

mransan commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +59,29 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+let number_of_int i : t = Obj.magic i
+
+let boolean b : t = Obj.magic (Js_boolean.to_js_boolean b)
```
I think we could provide the option to encode boolean as int but I still prefer to leave this API as is. The reason is that the caller might really want to put a `true` value in the JSON and it does not own the other side. 

Encoding bool as int decision is better left to the caller I think.

mransan commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +61,31 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+let objectArray a : t = Obj.magic a
+
+external stringify: 'a -> string = "JSON.stringify" [@@bs.val]
```
@bobzhang Should this be `t` rather `'a`. This would ensure a safer version.

bobzhang replied:
yes, being `t` is safe

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +61,31 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+let numberOfInt i : t = Obj.magic i
+
+let boolean b : t = Obj.magic (Js_boolean.to_js_boolean b)
```
you can encode ocaml boolean just as int,  no need for coercion, right?
I think for these `to_json` functions, the API should be symmetric, you need provide both, otherwise `encoding` does not make sense?

mransan replied:
@bobzhang In my case I really to have a boolean value in the JSON.

This module goal is to provide api to read and create any JSON values not just serializing OCaml values. JSON contains boolean we must provide a way to write a boolean. 

We could also provide an API like `boolAsInt` which would do what you describe. It's then up to the caller to decide what to use.

bobzhang replied:
ah, I see so your encoder is language agnostic. then should the API just be 
```
val boolean : Js.boolean -> Js.t 
```

mransan replied:
@bobzhang I'll update the PR, I also see the symetry problem you are talking about.

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +59,29 @@ let test (type a) (x : 'a) (v : a kind) : bool =
 external parse : string -> t = "JSON.parse" [@@bs.val]
 (* TODO: more docs when parse error happens or stringify non-stringfy value *)
+
+let null : t = Obj.magic Js.null 
```
maybe we can just say here
```
external null : t = "" [@@bs.val]
```

mransan commented:
@bobzhang I updated the PR based on your feedback. Indeed the use of external in the `.mli` guarantees that those functions are inlined!

Regarding the design around bool/boolean value:
* API is fixed and is now symmetric: `external boolean : Js.boolean -> t` is the symmetric of the 
  GADT case `|Boolean`
* Additionally I provide 2 functions for the OCaml `bool` type: `boolAsBoolean` and `boolAsInt`. The application can choose which one they prefer.

glennsl commented on jscomp/others/js_dict.mli:
```diff
@@ -0,0 +1,64 @@
+
+
+type 'a t
+(** Dictionary type (ie an '{ }' JS object). However it is restricted 
```
This comment will be associated with the module, just because it's the first doc comment in the module. You should add a doc comment before this so it'll pick that instead.

You also need to escape the `{` and `}` (and `[`, `]` and `@`).

I've had issues with having doc comments below the item they should be associated with. It should be possible, but i'm avoiding it because of such issues.

glennsl commented on jscomp/others/js_dict.mli:
```diff
@@ -0,0 +1,64 @@
+    when the existence of a key is certain. (i.e. when having called [keys]
+    function previously. 
+
+@example {|
```
`{| ... |}` is not valid doc comment syntax. It should be `{[ .. ]}` (square brackets instead of pipes). Why isn't consistent with OCaml you ask? Probably because they decided to make ocamldoc such a joy to use...

glennsl commented on jscomp/others/js_dict.mli:
```diff
@@ -0,0 +1,64 @@
+*)
+
+val exnGet : 'a t -> key -> 'a
+(** [exnGet dict key] returns the value if the [key] exists, otherwise
```
The first sentence of a doc comment should be a very short summary. This is used in indexes and by merlin and such.

glennsl commented on jscomp/others/js_dict.mli:
```diff
@@ -0,0 +1,64 @@
+
+val exnGet : 'a t -> key -> 'a
+(** [exnGet dict key] returns the value if the [key] exists, otherwise
+    @raise [Invalid_argument] exception *)
```
The first argument to the `@raise` tag needs to be a valid exception constructor. So you should just remove the brackets here.

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
+
+val reify_type : 'a -> 'b kind * 'b 
+  [@@ocaml.deprecated "Please use `reifyType`"]
+(** [reify_type v] returns both type and underlying value *) 
```
doc comments for deprecated items should include the @deprecated tag

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
+(** Those functions allows the construction of an arbitrary complex 
+    JSON values. 
+
+@example {|
```
Invalid code syntax again

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
+Js_dict.set dict "age" (Js_json.numberOfInt 30); 
+Js_dict.set dict "likes" 
+  (Js_json.stringArray [|"bucklescript";"ocaml";"js"|]);
+Js.log @@ Js_json.stringify (Js_json.object_ dict) 
```
These `@`s will need to be escaped too

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
 val test : 'a  -> 'b kind -> bool
+(** [test v kind] returns true if [v] is of [kind] *)
+
+(** {2 Construtors} *)
```
This isn't associated with any item, and will just be ignored by ocamldoc

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
+
+(** {2 Construtors} *)
+
+(** Those functions allows the construction of an arbitrary complex 
```
As will this

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
+external array_ : t array -> t = "%identity"
+(** Make a JSON array *)
+
+(** The functions below are specialized for specific array type which 
```
This will be ignored too.

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
+external objectArray : t Js_dict.t array -> t = "%identity"
+(** Make a JSON object array *)
+
+(** {2 String conversion} *)
```
As will this

glennsl commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
 external parse : string -> t = "JSON.parse" [@@bs.val]
+(** [parse s] returns JSON value 
+
+@example {|
```
And this has invalid code syntax again.

bobzhang replied:
can we rename it into `parse_exn`

bobzhang commented on jscomp/others/js_dict.ml:
```diff
@@ -27,6 +27,11 @@ type 'a t
+external unsafeGet : 'a t -> key -> 'a = "" [@@bs.get_index] 
+let exnGet dict key = 
+  match get dict key with
+  | None -> invalid_arg ("Js_dict.exnGet: missing key " ^ key)
```
why do we need this function here

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +61,35 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+external number : float -> t = "%identity"
+
+external numberOfInt : int -> t = "%identity"
```
`int` is not part of json standard, let's keep the API simple. you can always 
(float x) to convert it into float

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +61,35 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+external boolean : Js.boolean -> t = "%identity" 
+
+let boolAsBoolean b = (Js_boolean.to_js_boolean b) |> boolean 
```
should not be here, since this encoding is language agnostic

mransan replied:
Yes that one can be removed.

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +61,35 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+let boolAsBoolean b = (Js_boolean.to_js_boolean b) |> boolean 
+
+external boolAsInt : bool -> t = "%identity" 
```
should not be here

mransan replied:
see my global comment

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +61,35 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+external numberArray : float array -> t = "%identity"
+
+external intArray : int array -> t = "%identity"
```
same. since in JS  `int array` can be safely cast to `float array`, but I think such safe cast may find places in other module

bobzhang commented on jscomp/others/js_json.mli:
```diff
@@ -34,8 +38,106 @@ type _ kind =
+(** {2 String conversion} *)
 
 external parse : string -> t = "JSON.parse" [@@bs.val]
+(** [parse s] returns JSON value 
```
exnParse

mransan replied:
@bobzhang What exception type is raised ?

bobzhang commented on site/docsource/Dev-mode-How-to.adoc:
```diff
@@ -28,7 +28,7 @@ change `j.ml` (most likely you won't), so you probably don't need it
 [source,sh]
 -----------
-make bin/bsb.exe
+make -C bin bsb.exe
```
did you see my comments #1237
does that fix for you? @glennsl @mransan

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,6 +61,35 @@ let test (type a) (x : 'a) (v : a kind) : bool =
 (* TODO: more docs when parse error happens or stringify non-stringfy value *)
+
+let null : t = (Obj.magic (Js.null: 'a Js.null) : t)
+
```
```ocaml
external null : t = "" [@@bs.val]
```

mransan commented:
@bobzhang the reason I added those `array` api as well as the `int` in this module is to put in a single place **all the assumption about the runtime representation of OCaml value**.

An application cannot assume that for instance an `int array` can be casted to a `Js_json.t` value safely... only the compiler can. If I removed those functions then the application will be force to pay a performance penalty by calling:
* `Array.map` 
* `float_of_int`
Those calls are not inlined and can expensive and they can safely be avoided.

bobzhang commented:
I agree it makes sense to have it inside BuckleScript, maybe in a separate module, since it is generally useful, Js_FastCast ?

mransan commented:
@bobzhang great idea, would you like me to do it in this PR ? I would imagine a Js_cast is a good short name ?

mransan commented:
@bobzhang One thing though is the type that it cast to must be shared with the Js_json module right ?

bobzhang commented:
Yes, you only need cast between int and float, int array and float array etc. A separate PR would be nice

chenglou commented:
I'm not using `reify_type`, thanks though =)

mransan commented:
@bobzhang the latest commit address the following:
* remove `int` and `bool` APIs. The `Js_json` interface is now smaller and has the minimum set of constructor which are symmetric with the `kind` GADT 
* documentation is updated 
* merge with most recent master and latest @chenglou changes. 

Once this PR is merged I'll add `Js_cast` module.

bobzhang commented on jscomp/others/js_dict.mli:
```diff
@@ -0,0 +1,64 @@
+*)
+
+val exnGet : 'a t -> key -> 'a
+(** [exnGet dict key] returns the value if the [key] exists, otherwise
```
can we remove this API, this looks not very useful, exception based API is pervasive in OCaml mostly because it is efficient, but this characteristic is lost in JS land

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,7 +61,28 @@ let test (type a) (x : 'a) (v : a kind) : bool =
 external stringifyAny : 'a -> string option = "JSON.stringify" [@@bs.val] [@@bs.return undefined_to_opt]
 (* TODO: more docs when parse error happens or stringify non-stringfy value *)
+
+let null : t = (Obj.magic (Js.null: 'a Js.null) : t)
```
```ocaml
external null : t = "" [@@bs.obj]
```

mransan replied:
@bobzhang this does not work anymore, it's now an empty object not a `null` value.

mransan replied:
@bobzhang This unit test is now failing
```OCaml
let () = 
  let json = Js.Json.null |> Js.Json.stringify |> Js.Json.parse in 
  let ty, x = Js.Json.reifyType json in
  match ty with
  | Js.Json.Null -> true_ __LOC__
  | _ -> Js.log x; false_ __LOC__
```

mransan replied:
`[@@bs.val]` fixes it

bobzhang commented on jscomp/others/js_json.ml:
```diff
@@ -60,7 +61,28 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+external array_ : t array -> t = "%identity"
+
+external stringArray : string array -> t = "%identity"
```
strictly speaking maybe you don't need this, since you can cast string array , number array to t array as well, but this API looks intuitive I am good with it too

mransan commented:
@bobzhang latest commit addresses your comments

bobzhang commented:
Thanks!

chenglou commented on jscomp/others/js_json.ml:
```diff
@@ -60,7 +61,28 @@ let test (type a) (x : 'a) (v : a kind) : bool =
+
+external object_ : t Js_dict.t -> t = "%identity"
+
+external array_ : t array -> t = "%identity"
```
@bobzhang what's the convention? `_array` or `array_`? We've gone with the former because it corresponds to name mangling for obj
